contributors/design-proposals/auth/kubectl-exec-plugins.md

```diff
@@ -0,0 +1,193 @@
+# Out-of-tree client authentication providers
+
+Author: @ericchiang
+
+# Objective
+
+This document describes a credential rotation strategy for kubectl and client-go
+using an exec-based plugin mechanism.
+
+# Background
+
+Kubernetes clients can provide three kind of credentials: bearer tokens, TLS
+client certs, and basic auth username and password. Kubeconfigs can either
+in-line the credential, load credentials from a file, or can use an `AuthProvider`
+to actively fetch and rotate credentials. `AuthProviders` are compiled into kubectl
+and target specific providers (GCP, Keystone, Azure AD) or implement a
+specification supported but a subset of vendors (OpenID Connect).
+
+External tools that do rotation generally have to be called on a regular bases in
+case the current credentials have expired, which leads to wrapping `kubectl` with
```

Comment:
This is not required if the refresh token login is built into the plugin. For instance Azure Active Directory plugin does this without any external wrapping see: https://github.com/kubernetes/client-go/blob/1b75876d45719a84085528ee621939b0c68a4b85/plugin/pkg/client/auth/azure/azure.go#L276

Reply:
We would like to be able to support refreshing credentials for additional providers who do not have a plugin compiled in. It's not feasible to require every new provider or enterprise with custom tooling to submit a PR to kubernetes and we would like to discourage forking or wrapping kubectl in an alias as is often currently done. (not to say this proposal is not the only solution to this problem)